Engine/EffectInstanceRenderRoI.cpp

```diff
@@ -736,39 +728,44 @@ EffectInstance::renderRoI(const RenderRoIArgs & args,
     ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
     ////////////////////////////// Compute RoI depending on render scale ///////////////////////////////////////////////////
 
-
     RectI downscaledImageBoundsNc = rod.toPixelEnclosing(args.mipMapLevel, par);
     RectI upscaledImageBoundsNc = rod.toPixelEnclosing(0, par);
-    {
-        ///Make sure the RoI falls within the image bounds
-        ///Intersection will be in pixel coordinates
-        if (frameArgs->tilesSupported) {
-            if (renderFullScaleThenDownscale) {
-                if ( !roi.clipIfOverlaps(upscaledImageBoundsNc) ) {
-                    return eRenderRoIRetCodeOk;
-                }
-                assert(roi.x1 >= upscaledImageBoundsNc.x1 && roi.y1 >= upscaledImageBoundsNc.y1 &&
-                       roi.x2 <= upscaledImageBoundsNc.x2 && roi.y2 <= upscaledImageBoundsNc.y2);
-            } else {
-                if ( !roi.clipIfOverlaps(downscaledImageBoundsNc) ) {
-                    return eRenderRoIRetCodeOk;
-                }
-                assert(roi.x1 >= downscaledImageBoundsNc.x1 && roi.y1 >= downscaledImageBoundsNc.y1 &&
-                       roi.x2 <= downscaledImageBoundsNc.x2 && roi.y2 <= downscaledImageBoundsNc.y2);
-            }
-#ifndef NATRON_ALWAYS_ALLOCATE_FULL_IMAGE_BOUNDS
-            ///just allocate the roi
-            upscaledImageBoundsNc.clipIfOverlaps(roi);
-            downscaledImageBoundsNc.clipIfOverlaps(args.roi);
-#endif
+    RectI roi;
+
+    if (renderFullScaleThenDownscale) {
+        //We cache 'image', hence the RoI should be expressed in its coordinates
+        //renderRoIInternal should check the bitmap of 'image' and not downscaledImage!
+        roi = args.roi.toNewMipMapLevel(args.mipMapLevel, 0, par, rod);
+
+        if (frameArgs->tilesSupported && !roi.clipIfOverlaps(upscaledImageBoundsNc)) {
+            return eRenderRoIRetCodeOk;
         }
+        assert( upscaledImageBoundsNc.contains(roi));
+    } else {
+        roi = args.roi;
+
+        if (frameArgs->tilesSupported && !roi.clipIfOverlaps(downscaledImageBoundsNc)) {
+            return eRenderRoIRetCodeOk;
+        }
+        assert(downscaledImageBoundsNc.contains(roi));
```

Comment:
Attach any input to viewer (and/or scale) and assert fail.

Reply:
Fixed by https://github.com/NatronGitHub/Natron/pull/930 . I accidentally made the assert apply to the case where tiles are not supported, which does not match the original code. I've restored the old behavior  so the assert only applies when tiles are supported.